File: app/src/main/java/social/entourage/android/api/tape/EncounterTapeService.java
package social.entourage.android.api.tape;

import android.app.Service;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.net.ConnectivityManager;
import android.net.NetworkInfo;
import android.os.IBinder;

import com.squareup.otto.Subscribe;

import javax.inject.Inject;

import social.entourage.android.EntourageApplication;
import social.entourage.android.api.tape.Events.*;
import social.entourage.android.map.encounter.CreateEncounterPresenter.EncounterUploadCallback;
import social.entourage.android.map.encounter.CreateEncounterPresenter.EncounterUploadTask;
import social.entourage.android.tools.BusProvider;

public class EncounterTapeService extends Service implements EncounterUploadCallback {

    // ----------------------------------
    // ATTRIBUTES
    // ----------------------------------

    @Inject
    EncounterTapeTaskQueue queue;
    private static boolean running;
    private boolean connected;

    // ----------------------------------
    // LIFECYCLE
    // ----------------------------------

    @Override
    public void onCreate() {
        super.onCreate();
        EntourageApplication.get(this).getEntourageComponent().inject(this);
        connected = isConnected();
        BusProvider.getInstance().register(this);
    }

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        if (queue != null) {
            executeNext();
        } else {
            stopService();
        }
        return START_STICKY;
    }

    @Override
    public IBinder onBind(Intent intent) {
        return null;
    }

    // ----------------------------------
    // METHODS
    // ----------------------------------

    private boolean isConnected() {
        ConnectivityManager connectivityManager = (ConnectivityManager) this.getSystemService(Context.CONNECTIVITY_SERVICE);
        NetworkInfo activeNetInfo = connectivityManager.getActiveNetworkInfo();
        return activeNetInfo != null && activeNetInfo.isConnected();
    }

    private void stopService() {
        BusProvider.getInstance().unregister(this);
        stopSelf();
    }

    private void executeNext() {
        if (connected) {
            if (running) {
                return;
            }
            EncounterUploadTask task = queue.peek();
            if (task != null) {
                running = true;
                task.execute(this);
            } else {
                stopService();
            }
        }
    }

    @Override
    public void onSuccess() {
        running = false;
        queue.remove();
        executeNext();
    }

    @Override
    public void onFailure() {
        running = false;
        executeNext();
    }

    // ----------------------------------
    // BUS LISTENERS
    // ----------------------------------

    @Subscribe
    public void onConnectionChanged(OnConnectionChangedEvent event) {
        connected = event.isConnected();
        if (connected) executeNext();
    }

    // ----------------------------------
    // INNER CLASSES
    // ----------------------------------

    public static class ConnectionChangeReceiver extends BroadcastReceiver {

        @Override
        public void onReceive(Context context, Intent intent) {
            BusProvider.getInstance().register(this);
            ConnectivityManager connectivityManager = (ConnectivityManager) context.getSystemService(Context.CONNECTIVITY_SERVICE);
            NetworkInfo activeNetInfo = connectivityManager.getActiveNetworkInfo();
            if (activeNetInfo != null && activeNetInfo.isConnected()) {
                BusProvider.getInstance().post(new OnConnectionChangedEvent(true));
            } else {
                BusProvider.getInstance().post(new OnConnectionChangedEvent(false));
            }
            BusProvider.getInstance().unregister(this);
        }

    }
}

File: app/src/main/java/social/entourage/android/api/tape/EncounterTapeService.kt
package social.entourage.android.api.tape

import android.app.Service
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.net.ConnectivityManager
import android.net.NetworkCapabilities
import android.os.Build
import android.os.IBinder
import com.squareup.otto.Subscribe
import social.entourage.android.EntourageApplication.Companion.get
import social.entourage.android.api.model.tour.Encounter
import social.entourage.android.api.tape.EncounterTaskResult.OperationType
import social.entourage.android.api.tape.Events.OnConnectionChangedEvent
import social.entourage.android.tools.EntBus
import social.entourage.android.tour.encounter.EncounterUploadCallback
import javax.inject.Inject

class EncounterTapeService : Service(), EncounterUploadCallback {
    // ----------------------------------
    // ATTRIBUTES
    // ----------------------------------
    @JvmField
    @Inject
    var queue: EncounterTapeTaskQueue? = null
    private var connected = false

    // ----------------------------------
    // LIFECYCLE
    // ----------------------------------
    override fun onCreate() {
        super.onCreate()
        get(this).components.inject(this)
        connected = isConnected()
        EntBus.register(this)
    }

    override fun onStartCommand(intent: Intent, flags: Int, startId: Int): Int {
        if (queue != null) {
            executeNext()
        } else {
            stopService()
        }
        return START_STICKY
    }

    override fun onBind(intent: Intent): IBinder? {
        return null
    }

    // ----------------------------------
    // METHODS
    // ----------------------------------
    private fun isConnected(): Boolean {
        return Companion.isConnected(this)
    }

    private fun stopService() {
        EntBus.unregister(this)
        stopSelf()
    }

    private fun executeNext() {
        if (connected) {
            if (running) {
                return
            }
            queue?.peek()?.let {task ->
                running = true
                task.execute(this)
            } ?: run {
                stopService()
            }
        }
    }

    override fun onSuccess(encounter: Encounter?, operationType: OperationType?) {
        running = false
        queue?.remove()
        executeNext()
    }

    override fun onFailure(encounter: Encounter?, operationType: OperationType?) {
        running = false
        executeNext()
    }

    // ----------------------------------
    // BUS LISTENERS
    // ----------------------------------
    @Subscribe
    fun onConnectionChanged(event: OnConnectionChangedEvent) {
        connected = event.isConnected
        if (connected) executeNext()
    }

    // ----------------------------------
    // INNER CLASSES
    // ----------------------------------
    class ConnectionChangeReceiver : BroadcastReceiver() {
        override fun onReceive(context: Context, intent: Intent) {
            EntBus.register(this)
            val connectivityManager = context.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
            EntBus.post(OnConnectionChangedEvent(EncounterTapeService.isConnected(context)))
            EntBus.unregister(this)
        }
    }

    companion object {
        private var running = false
        private fun isConnected(context: Context): Boolean {
            val connectivityManager = context.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
            return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
                val networkCapabilities = connectivityManager.activeNetwork ?: return false
                val actNw =
                    connectivityManager.getNetworkCapabilities(networkCapabilities) ?: return false
                when {
                    actNw.hasTransport(NetworkCapabilities.TRANSPORT_WIFI) -> true
                    actNw.hasTransport(NetworkCapabilities.TRANSPORT_CELLULAR) -> true
                    actNw.hasTransport(NetworkCapabilities.TRANSPORT_ETHERNET) -> true
                    else -> false
                }
            } else {
                connectivityManager.run {
                    connectivityManager.activeNetworkInfo?.run {
                        when (type) {
                            ConnectivityManager.TYPE_WIFI -> true
                            ConnectivityManager.TYPE_MOBILE -> true
                            ConnectivityManager.TYPE_ETHERNET -> true
                            else -> false
                        }

                    }
                } ?: false
            }
        }

    }
}